### ISSUE TITLE
API seems to send initial data for each request

### ISSUE THREAD

**Issue by jarlrmai** [lucashenning/logstash-filter-rest#10]
Using the following filter, the API seems not to refresh the token on each pass

```
input {
  # Read all documents from Elasticsearch matching the given query
  elasticsearch {
    hosts => "localhost"
    index => "postcodes"
    scan => false
    size => 1
    scroll => "1m"
    #query => '{"from":0,"size":10,"query":{"match_all":{}}}'
    query => '{"query":{"range":{"id":{"gte":0,"lte":10}}}}'
    }
}

filter{

    rest{
      request => {
        url => "https://maps.googleapis.com/maps/api/distancematrix/json?"
        method => "get"

      params => {
        "units" => "imperial"
        "origins" => "%{postcode}"
        "destinations" => "AA11AA|BB11BB"
        "key" => "APIKEY"
        }
        }
      json => true
      sprintf => true
      target => "google_response"
    }

    mutate{
      add_field => { "json" => "%{google_response}"}
      remove_field => [ "google_response" ]
      #gsub => [ "json", "[\\]", "_" ]
    }

    json {
       source => "json"
       target => "json"
    }
 if [json][rows][0][elements[0][status] == "OK"
    {
    mutate{
      add_field => { "postymccode" => "%{postcode}" }
      add_field => { "origin" => "%{[json][origin_addresses][0]}" }
      add_field => { "distSDGH" => "%{[json][rows][0][elements[0][distance][text]}" }
      add_field => { "distODGH" => "%{[json][rows][0][elements[1][distance][text]}" }
      add_field => { "distSDGHmeters" => "%{[json][rows][0][elements[0][distance][value]}" }
      add_field => { "distODGHmeters" => "%{[json][rows][0][elements[1][distance][value]}" }

      add_field => { "timeSDGH" => "%{[json][rows][0][elements[0][duration][text]}" }
      add_field => { "timeODGH" => "%{[json][rows][0][elements[1][duration][text]}" }
      add_field => { "timeSDGHmins" => "%{[json][rows][0][elements[0][duration][value]}" }
      add_field => { "timeODGHmins" => "%{[json][rows][0][elements[1][duration][value]}" }

      remove_field => [ "json" ]
    }
    }
    else
    {

      drop {}

    }
}


output {
        elasticsearch {
         hosts => [ "localhost" ]

```

Output:

Settings: Default pipeline workers: 4
Pipeline main started
{
          "postcode" => "BB11BA",
                "id" => 6,
          "@version" => "1",
        "@timestamp" => "2016-09-21T08:19:33.688Z",
            "origin" => "Bristol Cl, Blackburn BB1 1BA, UK",
          "distSDGH" => "31.6 mi",
          "distODGH" => "34.1 mi",
    "distSDGHmeters" => 50814,
    "distODGHmeters" => 54917,
          "timeSDGH" => "56 mins",
          "timeODGH" => "43 mins",
      "timeSDGHmins" => 3386,
      "timeODGHmins" => 2588
}
{
          "postcode" => "BB11AD",
                "id" => 4,
          "@version" => "1",
        "@timestamp" => "2016-09-21T08:19:33.886Z",
            "origin" => "Bristol Cl, Blackburn BB1 1BA, UK",
          "distSDGH" => "31.6 mi",
          "distODGH" => "34.1 mi",
    "distSDGHmeters" => 50814,
    "distODGHmeters" => 54917,
          "timeSDGH" => "56 mins",
          "timeODGH" => "43 mins",
      "timeSDGHmins" => 3386,
      "timeODGHmins" => 2588
}

You can see the postcode changes on the second query but the info sent to the API (returned as origin) is the same postcode as the 1st lookup.

**Comment by gandalfb** [comment]
Hi,

this sounds like the bug fixed in one of the pull requests. #8

Could you please verify you have the latest gem installed v0.2.1?
https://rubygems.org/gems/logstash-filter-rest/versions/0.2.1

Thanks

**Comment by jarlrmai** [comment]
```
user@dcm4che:~/logstash-2.3.4$ bin/logstash-plugin list --verbose | grep rest
logstash-filter-rest (0.2.1)
user@dcm4che:~/logstash-2.3.4$
```

**Comment by jarlrmai** [comment]
The odd thing is I was convinced it was working fine when I 1st started getting results back from the API.

**Comment by gandalfb** [comment]
Odd thing for me is the same :-) It worked for me... so I guess I am missing something.  

Are you able to provide a minimal example for reproduction?
Maybe this is something with the google API.
E.g. you are using the '|' within parameters and the 'mutate' and 'json' config feels obsolete which adds a layer of confusion (to me) for better understanding. 

I will be able to double check sometime end of next week.

**Comment by JohnUSA** [comment]
I'm experiencing this issue too. When I pass data within 'params' it is sending the data from the initial event. But when I add data as part of the URL query string, it works.

**Comment by gandalfb** [comment]
At least it should work as @JohnUSA describes.

Tried it with this minimal example

``` bash
bin/logstash --debug -e 'input { stdin{} } filter { rest { request => { url => "http://jsonplaceholder.typicode.com/users/%{message}" } json => true sprintf => true } } output {stdout { codec => rubydebug }}'
```

The bug did not show due to the `rspec` always initializing a new instance and using `.freeze` not correctly.

The root cause seems to be a thing with object references. The request is an array holding request opts like `params`. The array is frozen and duplicated and then sprintf'ed.

The inlying opts are a hash which seems not to be duplicated and therefore only referenced to always the same object.

So the bug is my false assumption that a `.freeze` also freezes complex objects within the array or `.dup` also copies the complex object. Instead the array is duplicated but not the `params` and therefore overwritten with the first event loosing any original config.

Long story short there are three options:
- use your get params within the url
- wait for fix
- adapt the `rest.rb`

``` ruby
    #request = @request.dup
    request = Marshal.load(Marshal.dump(@request))
```

You may try this with

``` bash
bin/logstash --debug -e 'input { stdin{} } filter { rest { request => { url => "https://jsonplaceholder.typicode.com/posts" method => "get" params => { "userId" => "%{message}" } headers => { "Content-Type" => "application/json" } } json => true sprintf => true } } output {stdout { codec => rubydebug }}'
```

**Comment by JohnUSA** [comment]
@gandalfb fix #11 fixes it for me

**Comment by lucashenning** [comment]
Solved in v0.5.0 Thanks everybody
